File: src/main/java/dao/RepositoryDao.java
package dao;

import models.File;
import models.Repository;
import org.hibernate.NonUniqueResultException;
import org.hibernate.Session;
import org.hibernate.Transaction;
import utils.SessionFactoryUtil;

import javax.persistence.NoResultException;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import java.util.List;

public class RepositoryDao extends Dao<Repository> {

    private int userID;
    private String password;

    public RepositoryDao(int userID, String password) {
        this.userID = userID;
        this.password = password;
    }

    public Repository findById(int id) {
        return SessionFactoryUtil.getSessionFactory(userID, password).openSession().get(Repository.class, id);
    }

    public Repository findbyURL(String url) {
        Session session = SessionFactoryUtil.getSessionFactory(userID, password).openSession();
        CriteriaQuery<Repository> criteriaQuery = session.getCriteriaBuilder().createQuery(Repository.class);
        Root<Repository> root = criteriaQuery.from(Repository.class);
        criteriaQuery.select(root);
        criteriaQuery.where(
                session.getCriteriaBuilder().equal(root.get("url"), url)
        );

        Repository result = null;
        try {
            result = session.createQuery(criteriaQuery).getSingleResult();
        } catch (NonUniqueResultException e) {
            e.printStackTrace();
        } catch (NoResultException ignored) {}

        return result;
    }

    public void save(Repository repo) {
        Session session = SessionFactoryUtil.getSessionFactory(userID, password).openSession();
        Transaction tx = session.beginTransaction();
        session.save(repo);
        tx.commit();
        session.close();
    }

    public void update(Repository repo) {
        Session session = SessionFactoryUtil.getSessionFactory(userID, password).openSession();
        Transaction tx = session.beginTransaction();
        session.update(repo);
        tx.commit();
        session.close();
    }

    public void delete(Repository repo) {
        Session session = SessionFactoryUtil.getSessionFactory(userID, password).openSession();
        Transaction tx = session.beginTransaction();
        session.delete(repo);
        tx.commit();
        session.close();
    }

    public Package findPackageById(int id) {
        return SessionFactoryUtil.getSessionFactory(userID, password).openSession().get(Package.class, id);
    }

    public List<Repository> getAll() {
        return (List<Repository>)
                SessionFactoryUtil.getSessionFactory(userID, password).openSession()
                        .createQuery("from Repository").list();
    }

    
}

File: src/main/kotlin/dao/RepositoryDao.kt
package dao

import models.Repository
import org.hibernate.NonUniqueResultException
import utils.SessionFactoryUtil
import utils.UIUtil

import javax.persistence.NoResultException

class RepositoryDao(private val userID: Int, private val password: String) : Dao<Repository>() {

    override val all: List<Repository>?
        get() = try {
            SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession()
                    .createQuery("from Repository").list() as List<Repository>
        } catch (e: Exception) {
            when (e) {
                is org.hibernate.HibernateException, is javax.persistence.PersistenceException -> {
                    UIUtil.showHibernateError(e.toString())
                } else -> { e.printStackTrace() }
            }
            null
        }

    override fun findById(id: Int): Repository? {
        return try {
            SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession().get(Repository::class.java, id)
        } catch (e: Exception) {
            when (e) {
                is org.hibernate.HibernateException, is javax.persistence.PersistenceException -> {
                    UIUtil.showHibernateError(e.toString())
                } else -> { e.printStackTrace() }
            }
            null
        }
    }

    fun findByURL(url: String): Repository? {
        val session = SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession()
        val criteriaQuery = session.criteriaBuilder.createQuery(Repository::class.java)
        val root = criteriaQuery.from(Repository::class.java)
        criteriaQuery.select(root)
        criteriaQuery.where(
                session.criteriaBuilder.equal(root.get<Any>("url"), url)
        )

        var result: Repository? = null
        try {
            result = session.createQuery(criteriaQuery).singleResult
        } catch (e: NonUniqueResultException) {
            e.printStackTrace()
        }catch (e: Exception) {
            when (e) {
                is org.hibernate.HibernateException, is javax.persistence.PersistenceException -> {
                    UIUtil.showHibernateError(e.toString())
                } else -> { e.printStackTrace() }
            }
        } catch (ignored: NoResultException) {}

        return result
    }

    override fun save(obj: Repository) {
        val session = SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession()
        val tx = session.beginTransaction()
        session.save(obj)
        tx.commit()
        session.close()
    }

    override fun update(obj: Repository) {
        val session = SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession()
        val tx = session.beginTransaction()
        session.update(obj)
        tx.commit()
        session.close()
    }

    override fun delete(obj: Repository) {
        val session = SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession()
        val tx = session.beginTransaction()
        session.delete(obj)
        tx.commit()
        session.close()
    }

    fun findPackageById(id: Int): Package? {
        return try {
            SessionFactoryUtil.getSessionFactory(userID, password)!!.openSession().get(Package::class.java, id)
        } catch (e: Exception) {
            when (e) {
                is org.hibernate.HibernateException, is javax.persistence.PersistenceException -> {
                    UIUtil.showHibernateError(e.toString())
                } else -> { e.printStackTrace() }
            }
            null
        }
    }

}
